### ISSUE TITLE
TCH false positive with PEP-695 upper bound specification

### ISSUE THREAD

**Issue by bryanforbes** [astral-sh/ruff#7966]
Running the following command with Ruff `0.0.292`:

```
ruff --target-version py312 --select TCH --isolated --no-cache blah.py
```

Given the following code:

```python
from __future__ import annotations

from collections.abc import Callable
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from .foo import Record

type RecordOrThings = Record | int | str
type RecordCallback[R: Record] = Callable[[R], None]


def process_record[R: Record](record: R) -> None:
    ...


class RecordContainer[R: Record]:
    def add_record(self, record: R) -> None:
        ...
```

`TCH004` is reported on line 7, but all uses of `Record` are only for annotations.

**Comment by charliermarsh** [comment]
Does look like a bug.
